### ISSUE TITLE
[Feature request] Right Click variable to find in tree or a search function.

### ISSUE THREAD

**Issue by tarnos12** [adventuregamestudio/ags#1685]
**Describe the problem**
In larger projects finding a character/item/GUI might be a problem as there is no way to search the tree.
This can be somehow avoided by using folders, but if I need to find something specific that I am currently referencing in a script then I have to go through a big chunk of menus to find it(usually a problem when looking for GUI or an Item, even more so when looking for a GUIControl, you have to either remember which GUI holds that specific control or search through them all)

**Suggested change**
Right click a variable in a script should give us an option similar to "Navigate(in tree)" option.

Second possible way is to add a search bar that would search by name of items/GUI/GUIControl/Character/rooms/scripts etc.(like a filter)
This could include auto complete(like in script which is already aware that we have `cPlayer or Label1`)
Ideally it would work like a filter and hide things that do not match the search(might also include special characters for extra filtering options such as || which works as "or" so I can search for `cPlayer || gGui1 || myScript || iCup`)

The filter would be extra useful for cases where I have multiple states of an item such as "iCupEmpty, iCupCoffee, iCupTea, iCupJuice` I could search for `iCup` and get all of those in a list.

**Comment by tarnos12** [comment]
> In my opinion these are very different things. A find in tree that jumps and select imposes few changes to the tree ui, but the filter search has many implications.

I am referring to the already implemented feature which allows you to right click a tab of a script/character/item/gui to find it in a tree.
Ideally it would extend to the script itself, so I can right click a variable and search in tree.
So there are probably no changes needed to the tree ui.
I don't know how it's done(perhaps it is required to have the specific tab open/active for the navigation to work)

It's been implemented here: [Forum post](https://www.adventuregamestudio.co.uk/forums/?topic=46378.0)
https://i.imgur.com/4kJX7.png

**Comment by ericoporto** [comment]
> It's been implemented here: [Forum post](https://www.adventuregamestudio.co.uk/forums/?topic=46378.0)
> https://i.imgur.com/4kJX7.png

Thanks for bringing up this thread, I wasn't aware of that! 

But there's been a confusion, I meant the navigate in tree feature is of a much lesser scope than the project search, which can be done in many different ways. Here's an example of project search, with autocomplete, but without the filter part... Once filtering is added, than the questions I posted need answer during design to be able to be implemented.

![search_box](https://user-images.githubusercontent.com/2244442/173562970-917ad672-6148-448e-af78-c276274bcc8d.gif)

**Comment by tarnos12** [comment]
> But there's been a confusion, I meant the navigate in tree feature is of a much lesser scope than the project search, which can be done in many different ways.

I didn't know if I should be creating separate feature requests so I put them together.
One being easier and second for the future perhaps.

Since Navigate(in tree) is already implemented, I assumed that making it work with a script variable might be easier.

Btw, does your example work if you collapse the tree? Will it open up the tree to show what you are looking for?
What if the thing you are looking for is inside a folder or multiple nested folders?
If that works by default then it's pretty useful.
I assume it doesn't work with GUI controls?

Having it even with basic functionality is nice, adding more in the future is an option.
Since it already has auto complete then it's great.

**Comment by ericoporto** [comment]
> Since Navigate(in tree) is already implemented, I assumed that making it work with a script variable might be easier.

Yes, I assume it's easier and a lesser scope, but I haven't looked specifically into it.

---

I found and rebased my old project search [here, it's not good, use as a thinking experiment]( https://github.com/ericoporto/ags/tree/feature-project-search). Caveats:
- it's not possible to search for a folder.
- in case of ambiguous results, it jumps to the first, since it's using selection as means of highlighting.

I am sure I made a filter search too at some point in the past, but I could not find my branch on it...

> does your example work if you collapse the tree? Will it open up the tree to show what you are looking for?
  - yes and yes, it will expand as needed to reveal if it's hidden. and also scroll.
> What if the thing you are looking for is inside a folder or multiple nested folders?
  - no problem there, it will expand. 
> it doesn't work with GUI controls?
  - no because the GUI controls are not nodes in the project tree, this is what I meant with being specific on the content it should search on, but inside content will probably not be highlightable.

**Comment by ericoporto** [comment]
deleted/redacted comments and moved it to the other issue!

I think the better way for this feature is to use the "Go to definition", and then handle the case where it comes from the auto generated header.

https://github.com/adventuregamestudio/ags/blob/e2f5fdfae6ec19dd8a4e38554f79cb9fef644225/Editor/AGS.Editor/Panes/ScriptEditor.cs#L975-L978

**Comment by ericoporto** [comment]
After #2142 and #2148 are in, this should be closed (yes, there still will be room objects and GUIControl, but these are more complicated...).

**Comment by ivan-mogilko** [comment]
The original user's post contains this:

> usually a problem when looking for GUI or an Item, even more so when looking for a GUIControl, you have to either remember which GUI holds that specific control or search through them all

Meaning being able to jump to gui controls may be even more essential, as they are harder for a user to find than GUIs, for example.
I had a guess that jumping to gui control may be relatively easy to make, if you iterate all GUIs, find one, then do combined action of opening a pane and selecting a control. If you are not planning to do this in the nearest future, it's worth opening a dedicated ticket for that.

**Comment by ericoporto** [comment]
ok, made a tentative commit for GUI Controls -> https://github.com/adventuregamestudio/ags/pull/2148/commits/094ba603121187baa220da617660ab530dd6336f

In my tests it worked fine.
